### ISSUE TITLE
Setting Stage / Host from the task 

### ISSUE THREAD

**Issue by rylxes** [deployphp/deployer#2576]
Hello , 
Just got to know about this project, it makes my life easier , its awesome !!!

i think it would be a good idea to set a task's stage or host from the config , so you do not need to add host when running the task i.e dep {{task}} .

**Comment by Schrank** [comment]
@rylxes thanks for the compliment! Can you a little elaborate? I don't understand you. Maybe some (pseudo) code or examples?

**Comment by rylxes** [comment]
thanks ,

something like this 

task(‘do:task’, function () {
  })->withStage($stage);

Where $stage is a string , … and we invoke the task with  ‘dep do:task’ , ... which would tell the task to run the application with stage $stage


i hope i am able to explain well

**Comment by Schrank** [comment]
If I'm not mistaken this should work with $task->select() and/or $task->addSelector() but I don't know how it works. @antonmedv do you have an example how to use it? select() is the 7.0 implementation for https://deployer.org/docs/tasks.html#filtering is it?

**Comment by antonmedv** [comment]
I’m still need to find time to write new docs for v7. Will add select docs.

**Comment by yakimov** [comment]
Ohh, I need that one too. Coz it's difficult to understand how to run prod, dev, staging tasks if I use the YAML configuration. 
If I'm right, I have to use labels instead of roles, stage, host... 

btw @lorisleiva  has written [the awesome article](https://lorisleiva.com/deploy-your-laravel-app-from-scratch) about how to use deployer v7.  I've got many tricks from that. Perhaps he could help with the documentation.

For me, it would be enough to take a look at an example deploy.yaml with these filters... 
@antonmedv would you public examples of deploy.yaml that show secrets of phpdeployer v7

**Comment by Schrank** [comment]
Leave this here. @lorisleiva I LOVE YOU! <3<3 Thanks for taking the time to make it clear how this works! 
https://lorisleiva.com/deploy-your-laravel-app-from-scratch/create-your-own-deployer-recipes#labelling-hosts-and-filtering-tasks

**Comment by Schrank** [comment]
TL;DR:
host have labels, e.g.

```php
host('example.org')
    ->set('labels', [
        'stage' => 'stage',
        'type' => ['db', 'web', 'something']
    ]);
```

And then you can set certain tasks to only run on these labels, e.g.

```php
task('deploy:symlinks')->select('type=web'); // already defined tasks, the task is only filtered but NOT changed
task('something:new', function() {
    echo 'do something';
})->select('stage=stage,type=something');
```

**Comment by yakimov** [comment]
Unfortunately, this doesn't work for YAML configuration (

`config:
  application:  'app'
  remote_user:  'deployer'
  repository:   'git@bitbucket.org:wsp/app.git'
  branch:       'master'
  shared_dirs:
    - uploads
  writable_dirs:
    - log
    - cache

hosts:
  app-prod-host1:
    hostname:    app-host1
    deploy_path: '/var/www/{{application}}.ru'
    labels:
      stage: prod
  app-prod-host2:
    hostname:    app-host2
    deploy_path: '/var/www/{{application}}.ru'
    labels:
      stage: prod
  app-stage:
    hostname:    app-host3
    deploy_path: '/var/www/{{application}}.ru'
    labels:
      stage: staging

tasks:
  deploy:
    - deploy:prepare
    - deploy:vendors
    - app:env
    - app:build
    - app:setup
    - app:migrate-db
    - app:clear-cache
    - deploy:publish
  app:clear-cache:
    script: 'cd {{release_or_current_path}} && make clear-cache'
  app:build:
    script: 'cd {{release_or_current_path}} && make build-frontend'
  app:setup:
    script: 'cd {{release_or_current_path}} && make setup-app'
  # I need to launch app:migrate-db only for stage=prod
  # On the staging host this isn't needed
  # cmd: dep deploy 'stage=prod' 
  app:migrate-db:
    script: 'cd {{release_or_current_path}} && make migrate-db'
    select: stage=prod
after:
  deploy:failed: deploy:unlock
`

**Comment by Schrank** [comment]
@yakimov I have no clue about the Yaml configuration, but you can set a breakpoint into \Deployer\Executor\Master::run and into \Deployer\Task\Task::select and check what your configuration is doing. In theory select() should get stage=prod and the run should get an array ['=', 'stage', 'prod']

**Comment by yakimov** [comment]
@Schrank yes, you are right. I wanted to do it the nearest weekend.
